sdk/core/Azure.Core/src/Shared/AzureEventSource.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+
+namespace Azure.Core.Diagnostics
+{
+    internal abstract class AzureEventSource: EventSource
+    {
+        private static string[] MainEventSourceTraits =
+        {
+            AzureEventSourceListener.TraitName,
+            AzureEventSourceListener.TraitValue
+        };
+
+        protected AzureEventSource(string eventSourceName): base(
+            DeduplicateName(eventSourceName),
+            EventSourceSettings.Default,
+            MainEventSourceTraits
+        )
+        {
+        }
+
+        private static string DeduplicateName(string eventSourceName)
+        {
+            HashSet<string> namesInUse = new();
+            foreach (var source in GetSources())
+            {
+                namesInUse.Add(source.Name);
+            }
+
+            if (!namesInUse.Contains(eventSourceName))
+            {
+                return eventSourceName;
+            }
```

Comment:
Is there a race condition here that would result in two registrations calling GetSources before either adds themself to the source list?

Reply:
Yeah, the problem is I didn't find anything reasonable I can synchronize on. Any ideas?

Reply:
Would using a _syncLock work?

Reply:
No.

1. this is shared code file so every assembly would get it's own sync object.
2. In every version of the assembly this would be a distinct type so they would get their own sync objects as well.

Reply:
I see.. Well it would be the same process, right? Could we use a Mutex - https://docs.microsoft.com/en-us/dotnet/api/system.threading.mutex?redirectedfrom=MSDN&view=net-5.0

Reply:
> Well it would be the same process, right? 

Yeah.

> (now deleted mutex comment)

That's actually a good idea. I'll look into named mutexes.

>Could we sync on something that is part of the Environment or process?

Yeah, the question is what.

Reply:
Looked into named mutexes, they are not very reliable cross plat (implemented using file locks on mac). I don't want to put code like that on the very important startup path.

Reply:
You could suffix a new `Guid` for these anyway rare cases. :)